File: pset6/readability/readability.py

def count_letters(s) -> int:
    '''counts the alphanum letters in given string'''
    cnt = 0
    for c in s:
        if c.isalnum():
            cnt += 1
    return cnt


def count_words(s) -> int:
    '''counts the words in given str'''
    cnt = 0
    for c in s:
        if c == " " or c == "\t" or c == "\n" or c == "\0":
            cnt += 1
    return cnt + 1


def count_sentences(s) -> int:
    '''counts the sentences in given string'''
    cnt = 0
    for i in range(len(s)):
        c = s[i]
        if c == "." or c == "!" or c == "?":
            if c == "." and i + 1 < len(s) and s[i + 1] == ".":
                continue
            cnt += 1
    return cnt


def main():
    print("Text: ")
    text = input()
    # getting all key values
    letters = float(count_letters(text))
    words = float(count_words(text))
    sentences = float(count_sentences(text))

    # calculating L and S
    L = (letters / words) * 100
    S = (sentences / words) * 100
    # calculating the index
    index = 0.0588 * L - 0.296 * S - 15.8

    if index > 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade: {round(index)}")


# checking input function
if __name__ == "__main__":
    main()

File: pset2/readability/readability.c
#include <ctype.h>
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

int count_letters(string s);
int count_words(string s);
int count_sentences(string s);


int main(void)
{
    string text = get_string("Text: ");

    // counting letters in text
    float letters = count_letters(text);

    //counting words in text
    float words = count_words(text);

    //counting sentences
    float sen = count_sentences(text);

    //calculating average number of letters per 100 words
    float L = (letters / words) * 100;
    //calculating average number of sentences per 100 words
    float S = (sen / words) * 100;
    //calculating index via Coleman-Liau index
    float index = 0.0588 * L - 0.296 * S - 15.8;

    //presenting the result
    if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", (int)round(index));
    }

}

//counts the alphabetical letters
int count_letters(string s)
{
    int res = 0;
    for (int i = 0; i < strlen(s); i++)
    {
        if (isalnum(s[i]))
        {
            res++;
        }
    }
    return res;
}

// counts the words in a given string, word can be ended by [" ", "\t", "\n", "\0"]
// pre increment the result to get even the last word
int count_words(string s)
{
    int res = 0;
    for (int i = 0; i < strlen(s); i++)
    {
        if (s[i] == ' ' || s[i] == '\t' || s[i] == '\n' || s[i] == '\0')
        {
            res++;
        }
    }
    return ++res;
}

// counts the sentences in the given text
int count_sentences(string s)
{
    int res = 0;
    for (int i = 0; i < strlen(s); i++)
    {
        if (s[i] == '.' || s[i] == '!' || s[i] == '?')
        {
            if (s[i] == '.' && i + 1 < strlen(s) && s[i + 1] == '.')
            {
                continue;
            }
            res++;
        }
    }
    return res;
}